src/NServiceBus.AzureStoragePersistence/SagaPersisters/Azure/AzureSagaPersister.cs

```diff
@@ -275,7 +280,7 @@ object ToEntity(Type entityType, DictionaryTableEntity entity)
         /// <param name="context">The current context.</param>
         public Task Save(IContainSagaData sagaData, SagaCorrelationProperty correlationProperty, SynchronizedStorageSession session, ContextBag context)
         {
-            return Persist(sagaData);
+            return Task.WhenAll(new [] { secondaryIndeces.Insert(sagaData, correlationProperty), Persist(sagaData) });
```

Comment:
You can't `Task.WhenAll` this two operations. The reason for storing  the secondary index first is the need to ensure, that we won't create two sagas with the same correlation id, hence keeping these two in the same order as they were (but awaited) is the way to go.